pennylane_lightning_gpu/src/tests/mpi/Test_MPIManager.cpp

```diff
@@ -0,0 +1,220 @@
+
+    SECTION("Apply allgather vector") {
+        std::vector<cp_t> sendBuf(1, {static_cast<PrecisionT>(rank), 0});
+        auto recvBuf = mpi_manager.allgather<cp_t>(sendBuf);
+
+        for (size_t i = 0; i < recvBuf.size(); i++) {
+            CHECK(recvBuf[i].real() == static_cast<PrecisionT>(i));
+            CHECK(recvBuf[i].imag() == static_cast<PrecisionT>(0));
+        }
+    }
+}
+
+TEMPLATE_TEST_CASE("MPIManager::Allreduce", "[MPIManager]", float, double) {
+    using PrecisionT = TestType;
+    using cp_t = std::complex<PrecisionT>;
+
+    MPIManager mpi_manager(MPI_COMM_WORLD);
+    int rank = mpi_manager.getRank();
+    int size = mpi_manager.getSize();
+
+    SECTION("Apply Allreduce scalar") {
+        cp_t sendBuf = {static_cast<PrecisionT>(rank), 0};
+        cp_t recvBuf;
+
+        mpi_manager.Allreduce<cp_t>(sendBuf, recvBuf, "sum");
+        CHECK(recvBuf.real() == static_cast<PrecisionT>((size - 1) * size / 2));
+        CHECK(recvBuf.imag() == static_cast<PrecisionT>(0));
+    }
+
+    SECTION("Apply allreduce scalar") {
+        cp_t sendBuf = {static_cast<PrecisionT>(rank), 0};
+        auto recvBuf = mpi_manager.allreduce<cp_t>(sendBuf, "sum");
+
+        CHECK(recvBuf.real() == static_cast<PrecisionT>((size - 1) * size / 2));
+        CHECK(recvBuf.imag() == static_cast<PrecisionT>(0));
+    }
+
+    SECTION("Apply Allreduce vector") {
+        std::vector<cp_t> sendBuf(1, {static_cast<PrecisionT>(rank), 0});
+        std::vector<cp_t> recvBuf(1);
+
+        mpi_manager.Allreduce<cp_t>(sendBuf, recvBuf, "sum");
+
+        CHECK(recvBuf[0].real() ==
+              static_cast<PrecisionT>((size - 1) * size / 2));
+        CHECK(recvBuf[0].imag() == static_cast<PrecisionT>(0));
+    }
+
+    SECTION("Apply allreduce vector") {
+        std::vector<cp_t> sendBuf(1, {static_cast<PrecisionT>(rank), 0});
+        auto recvBuf = mpi_manager.allreduce<cp_t>(sendBuf, "sum");
+
+        CHECK(recvBuf[0].real() ==
+              static_cast<PrecisionT>((size - 1) * size / 2));
+        CHECK(recvBuf[0].imag() == static_cast<PrecisionT>(0));
+    }
+}
+
+TEMPLATE_TEST_CASE("MPIManager::Bcast", "[MPIManager]", float, double) {
+    using PrecisionT = TestType;
+    using cp_t = std::complex<PrecisionT>;
+
+    MPIManager mpi_manager(MPI_COMM_WORLD);
+    int rank = mpi_manager.getRank();
+
+    SECTION("Apply Bcast scalar") {
+        cp_t sendBuf = {static_cast<PrecisionT>(rank), 0};
+        mpi_manager.Bcast<cp_t>(sendBuf, 0);
+        CHECK(sendBuf.real() == static_cast<PrecisionT>(0));
+        CHECK(sendBuf.imag() == static_cast<PrecisionT>(0));
+    }
+
+    SECTION("Apply Bcast vector") {
+        std::vector<cp_t> sendBuf(1, {static_cast<PrecisionT>(rank), 0});
+        mpi_manager.Bcast<cp_t>(sendBuf, 0);
+        CHECK(sendBuf[0].real() == static_cast<PrecisionT>(0));
+        CHECK(sendBuf[0].imag() == static_cast<PrecisionT>(0));
+    }
+}
+
+TEMPLATE_TEST_CASE("MPIManager::Sendrecv", "[MPIManager]", float, double) {
```

Comment:
Just wondering if we need to worry about potential send/recv issues here causing deadlock. Is it better practice to assume the sends are async and recv are sync? I'll happily defer to you or @vincentmr on this one.

Reply:
Good point! `MPI_Sendrecv` does not suffer from the circular deadlock problems of `MPI_Send` and `MPI_Recv` since `MPI_Sendrecv` combines the functionality of sending and receiving messages in a single call.